### ISSUE TITLE
AddContainer does not have support for docker network

### ISSUE THREAD

**Issue by thijscornelis** [dotnet/aspire#850]
Hi,

Is there some roadmap for adding more docker container capabilities? Right now, I see that there is support for Volumes and Replicas using the ResourceAnnotations. 

```
builder.AddContainer(..)
    .WithAnnotation(new VolumeMountAnnotation(..))
    .WithAnnotation(new ReplicaAnnotation(..))
```

I was looking to create a set of containers that use a docker network.  If it's something that is not meant to be done using ResourceAnnotations, could you guide me on how we should do it? I see no option to add arguments to the 'docker run' command? 

Greetings!

P.S. I'm loving this project already :)

**Comment by karolz-ms** [comment]
What is the benefit of using Docker network for your app?

**Comment by davidfowl** [comment]
We shouldn't add support for docker networks but we should make a default network so that containers can reference each other by host name.

**Comment by thijscornelis** [comment]
> What is the benefit of using Docker network for your app?

I was following a tutorial to get an Elastic container up & running. Didn't think twice if I really needed a separate docker network, as I'm quite new to containers.

**Comment by davidfowl** [comment]
That makes sense @thijscornelis. Can you share the tutorial?

**Comment by thijscornelis** [comment]
> That makes sense @thijscornelis. Can you share the tutorial?

Sure, here you go: https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html

**Comment by Blind-Striker** [comment]
Another potential approach could be to leverage Dotnet Test containers instead of using Aspire's custom Docker wrapper. Testcontainers is a widely-used and well-established library within the community, offering robust and tested functionalities.

If this aligns with the project goals, I'm ready to open an issue to explore integrating Testcontainers. Thoughts?

[Dotnet Testcontainers GitHub](https://github.com/testcontainers/testcontainers-dotnet)

**Comment by karolz-ms** [comment]
I agree with @davidfowl, we should be using a dedicated Docker network for running an Aspire app (and also use the Aspire component names for container names, instead of Docker auto-generated names). 

@Blind-Striker thank you for pointing out the Testcontainers project, looks interesting, I will take a closer look.

**Comment by luizcarlosfaria** [comment]
> We shouldn't add support for docker networks but we should make a default network so that containers can reference each other by host name.

### Default networks in docker work's some different.

We can see on Docker Docs "_Containers that attach to a custom network use Docker's embedded DNS server. The embedded DNS server forwards external DNS lookups to the DNS servers configured on the host._" at https://docs.docker.com/network/#dns-services

Service discovery is part of custom docker networks, and when we connect some containers at this network, we're expecting that some containers talk to each other. 

**Kafka** and **zookeper**, **etcd** and other service tools, **elk stack** or anything else, like a **kong**, **nginx** **envoy** etc need a custom docker network to use a docker service discovery.

**Comment by 3GDXC** [comment]
IMHO Network configuration would be good as it would allow true infrastructure as code and follow what best practices with regards layer separation and firewall/port configuration testing in the developer environments.

**Comment by davidfowl** [comment]
> IMHO Network configuration would be good as it would allow true infrastructure as code and follow what best practices with regards layer separation and firewall/port configuration testing in the developer environments.

Can you expand on this? What does that mean exactly?

**Comment by 3GDXC** [comment]
> > IMHO Network configuration would be good as it would allow true infrastructure as code and follow what best practices with regards layer separation and firewall/port configuration testing in the developer environments.
> 
> Can you expand on this? What does that mean exactly?

firstly, happy new year @davidfowl and team; hope you all had a good holiday period.
and apologies for the cryptic description above.

We have a number guiding principles from our security and infrastructure teams; that we are required to follow when designing application/system architecture.  These principles are all around the topic of secure by design; and state we should separating application from it's services (not owned by the application its self) at a network layer especially common/shared services like caching/messaging/database.  The separation may be provided by the network via vlan routing, firewall or reverse-proxy allowing traffic between groups of services by port mappings.

To support the secure by design mandated by our security & infrastructure teams have tooling that is used map network designs (visual) or reverse engineer (into artifacts) like powershell DSC which are used to check the networks/volumes etc. and alert or revert changes if any unauthorised change is detected.

I understand as part of the teams tooling k8, nginx, v-sphere, hyper-v is supported.

If aspire supported network configurations this would allow the continued tooling to work; I also understand that the teams for a number of customer account use Azure Arc in a multi-tenanted type scenario where a tenant is a department of the larger customer account.

I hope that give you a better understanding of where and why aspire network configuration would/could help.  I'd be happy to outline more if needed via DM

**Comment by davidfowl** [comment]
This seems unrelated to supporting docker networks, but maybe I'm missing something.

**Comment by ejsmith** [comment]
@davidfowl the test container stuff is really nice. What is the story with Aspire supporting integration tests?

**Comment by ejsmith** [comment]
Just found a sample of using Aspire for tests: https://github.com/dotnet/eShop/blob/main/tests/Ordering.FunctionalTests/OrderingApiFixture.cs

**Comment by DamianEdwards** [comment]
@ejsmith better support for integration testing is something we're tracking in #1704

**Comment by joperezr** [comment]
FYI @mitchdenny for thoughts around integration with AppModel

**Comment by mitchdenny** [comment]
Not something we are tackling for GA at this point.

**Comment by 3GDXC** [comment]
> We shouldn't add support for docker networks but we should make a default network so that containers can reference each other by host name.

I'm interested as to why network support is not getting considered? can you enlighten us? maybe the community could do some support if it's simply not a priority?

**Comment by davidfowl** [comment]
We're not sure what to implement exactly. If you go directly through the docker network, then you circumvent the networking of our orchestrator which means it isn't modeled and doesn't show up anywhere. Projects don't run as containers so you can't attach projects to a container network. Right now, container to container networking is via the host and the orchestrator (dcp) manages proxies to make things like replicas work (see https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview for more details). Service discovery might just work but we haven't considered what it means to build the end to end such that if you turned your project into a container and added it to the app model that service discovery would continue to work as is (the follow on implications of those changes etc).

I think we will need to support this, but there has to be experimentation in how to expose this in the app model. I believe there's support for it in the underlying orchestrator, but it's not exposed.

TL;DR: We want to do this, but have decided this isn't a priority for GA so no time has been spent exploring or fleshing out potential solutions.

**Comment by EwoutdBoer** [comment]
Would be nice to have Elastc search as a component. Then you don't need to use the Elastic tutorial and you also don't run in the 'separate network' issue

**Comment by ejsmith** [comment]
I started working on Elasticsearch support here: https://github.com/exceptionless/Exceptionless/blob/feature/aspire/src/Exceptionless.AppHost/ElasticsearchExtensions.cs

**Comment by davidfowl** [comment]
Moving this up the priority list

**Comment by mitchdenny** [comment]
I think we need to start getting concrete about what features we want here and what the benefits are going to be both in terms of local development, and what it means to deployment scenarios.

Are we just looking to provide a mechanism to fine tune the Docker network configuration? Or do we want to have a fully fletched network resource that we put in the manifest which deployment tools then need to interpret.

I think we are going to run into a translation problem between the network topology that we can represent in Aspire locally and what the deployment tools can represent.

For example, lets say that we want to have multiple Docker networks defined and used in a single Aspire. API TBD, but it might look something like this:

```csharp
var net1 = builder.AddNetwork("net1");
var net2 = builder.AddNetwork("net2");

var app1 = builder.AddContainer("app1", "image1")
                  .AttachTo(net1);

var app2 = builder.AddContainer("app2", "image2")
                  .AttachTo(net1)
                  .AttachTo(net2);

var app2 = builder.AddContainer("app3", "image3")
                  .AttachTo(net2);
```

This would result in DCP creating two networks, and when we attach the containers to them. But what is the benefit here for local development?

The above is also problematic for deployment scenarios because target deployment environments won't always be able to faithfully represent a complex network topology like this. For example when we deploy with AZD we deploy all the containers on the same container app environment. Would using networks cause us to create multiple container app environments to ensure network isolation?

... we couldn't represent a container being on two networks so I guess that would have to be an error.

**Comment by davidfowl** [comment]
Can we just create a singlet network for all containers? We want to go through the dcp proxy for them as well for container  -> host comms

**Comment by mitchdenny** [comment]
I'm fine with that. I'm just trying to draw the perimeter on what me mean when we say that we support Docker networks ;)

**Comment by mitchdenny** [comment]
Moving to backlog for now. We need to do some more design around this and understand the implications in the context of developer edge to cloud as well.

**Comment by JonathanAtCenterEdge** [comment]
Wanted to chime in as an engineer researching what migration would look like for our team,

Our developers have been utilizing docker-compose for the last 5-ish years but we're desiring to migrate to Aspire. We use an external docker-compose network so that our docker-compose split into different repositories can all communicate to each other.

Migration for us from `docker-compose` to Aspire would require the ability to spin up containers in Aspire with the ability to overload the network until we have all of our services on Aspire, only then could we switch to a theoretical "global" Aspire docker network.

**Comment by davidfowl** [comment]
We've made a change to the upcoming version of aspire to support container to container networking with the default aspire-network. This isn't extensible as we're working towards a bigger network modeling change.

**Comment by 3GDXC** [comment]
> We've made a change to the upcoming version of aspire to support container to container networking with the default aspire-network (#5628). This isn't extensible as we're working towards a bigger network modeling change.
> 
> This should help with migration from docker-compose and it will also improve building resources where multiple containers talk to each other.

that's fantastic to here; is there any visibility of the bigger network modeling change? IMHO, it would be very interesting to have the team do rounds of open standup/discussion like the early day of .NET CORE ;) always keen to here background discussion about direction and choice that need to be agreed/disagreed etc.

**Comment by JobaDiniz** [comment]
> Our developers have been utilizing docker-compose for the last 5-ish years but we're desiring to migrate to Aspire. We use an external docker-compose network so that our docker-compose files that are specific to each of our services that reside in individual service-specific repositories can all communicate to each other.

this is exactly my case @JonathanAtCenterEdge , did you figure out a path forward?

**Comment by paulomorgado** [comment]
@JobaDiniz how are you doing it with docker-compose? Are you generating the container images and then run docker-compose?

You can use containers with Aspire. I have configurations on some of my hosts to use the project, a container image, or a remote service. It's a pain to do that, but as soon as it's done it's just changing a bit of config in the Aspire host to switch modes.

**Comment by mhwtx** [comment]
Is there any way to modify the configuration of the session-scoped bridge network that Aspire creates? Just experienced an interesting scenario where I had some service calls failing because the Aspire bridge network has the default MTU of 1500, but I needed a lower MTU (1425) to get things working.

I have that lower MTU of 1425 set in my docker daemon json, and the default bridge network seems to respect this. However, the Aspire bridge networks do not.

**Comment by JonathanAtCenterEdge** [comment]
> > Our developers have been utilizing docker-compose for the last 5-ish years but we're desiring to migrate to Aspire. We use an external docker-compose network so that our docker-compose files that are specific to each of our services that reside in individual service-specific repositories can all communicate to each other.
> 
> this is exactly my case @JonathanAtCenterEdge , did you figure out a path forward?

@JobaDiniz We did not, we've been patiently waiting for more flexibility with Aspire (such as custom docker networks) before making the Aspire plunge.

**Comment by romanimm** [comment]
Maybe I am wrong, but this looks very similar to https://github.com/dotnet/aspire/pull/10710 
(@davidfowl can maybe delegate this to @copilot)

If only the generated docker-compose is relevant, this works for me.

```cs
public sealed class DockerNetworkAnnotation(string name) : IResourceAnnotation
{
    /// <summary>
    /// Gets the name of the Docker network.
    /// </summary>
    public string Name { get; } = name;
}
public static class DockerNetworkExtensions
{
    /// <summary>
    /// Adds a Docker network to the resource.
    /// </summary>
    /// <typeparam name="T">The type of the resource.</typeparam>
    /// <param name="builder">The resource builder.</param>
    /// <param name="name">The name of the Docker network.</param>
    /// <returns>The resource builder.</returns>
    public static IResourceBuilder<T> WithNetwork<T>(this IResourceBuilder<T> builder, string name)
        where T : IResource
    {
        builder.WithAnnotation(new DockerNetworkAnnotation(name), ResourceAnnotationMutationBehavior.Append);

        return builder.WithManifestPublishingCallback(context =>
        {
            var annotations = builder.Resource.Annotations.OfType<DockerNetworkAnnotation>().ToList();
            if (annotations.Count != 0)
            {
                context.Writer.WriteStartArray("networks");
                foreach (var annotation in annotations)
                {
                    context.Writer.WriteStringValue(annotation.Name);
                }
                context.Writer.WriteEndArray();
            }
        });
    }
}
```
